medschool/src/parse.rs

```diff
@@ -0,0 +1,286 @@
+            // we have to go deeper!
+            syn::GenericArgument::Type(Type::Path(generic_path)) => {
+                // that's it, we've reached the final type
+                inner_type = match generic_path.path.segments.last() {
+                    Some(t) => Some(t.ident.clone()),
+                    None => break,
+                };
+
+                current_argument = generic_path.path.segments.last()?.arguments.clone();
+            }
+            _ => break,
+        }
+    }
+
+    inner_type
+}
+
+/// Converts a [`syn::ItemMod`] into a [`PartialType`].
+fn parse_item_mod(item_mod: syn::ItemMod) -> Option<PartialType> {
+    let thing_docs_untreated = docs_from_attributes(item_mod.attrs);
+    Some(PartialType {
+        ident: item_mod.ident.to_string(),
+        docs: thing_docs_untreated,
+        fields: Default::default(),
+    })
+}
+
+/// Converts a [`syn::ItemEnum`] into a [`PartialType`].
+fn parse_item_enum(item: syn::ItemEnum) -> Option<PartialType> {
+    let thing_docs_untreated = docs_from_attributes(item.attrs);
+    let is_internal = thing_docs_untreated
+        .iter()
+        .any(|doc| doc.contains(r"<!--${internal}-->"));
+
+    // We only care about types that have docs.
+    (!thing_docs_untreated.is_empty() && !is_internal).then(|| PartialType {
+        ident: item.ident.to_string(),
+        docs: thing_docs_untreated,
+        fields: Default::default(),
+    })
+}
+
+/// Converts a [`syn::ItemStruct`] into a [`PartialType`].
+fn parse_item_struct(item: syn::ItemStruct) -> Option<PartialType> {
+    let mut docs = docs_from_attributes(item.attrs);
+
+    if docs.is_empty() {
+        return None;
+    }
+
+    for doc in docs.iter_mut() {
+        // removes docs that we don't want in `configuration.md`
+        if doc.contains(r"<!--${internal}-->") {
+            return None;
+        }
+
+        // `trim` is too aggressive, we just want to remove 1 whitespace
+        if doc.starts_with(' ') {
+            doc.remove(0);
+        }
+    }
+
+    docs.push("\n".to_string());
+
+    // we used to remove any duplicate fields such as two fields with the same
+    // type by converting them to a HashSet
+    // for example, struct {a : B, b: B} would duplicate docs for B
+    // for our use case this is not necessary, and somehow ends up dropping
+    // fields so we're just going to keep the fields as
+    // they are and consider this again later
+    let fields = item
+        .fields
+        .into_iter()
+        .filter_map(PartialField::new)
+        .collect::<BTreeSet<_>>();
+
+    // We only care about types that have docs.
+    (!fields.is_empty()).then(|| PartialType {
+        ident: item.ident.to_string(),
+        docs,
+        fields,
+    })
+}
+
+/// Converts a [`syn::Item`] into a [`PartialType`], if possible.
+#[tracing::instrument(level = "trace", ret)]
+fn parse_syn_item_into_partial_type(item: syn::Item) -> Option<PartialType> {
```

Comment:
Another good candidate for becoming a `From<syn::Item> for PartialType`, we're dealing with `Option`s in some of these cases, so this would have to be `TryFrom`, and we `.ok()` or ignore the errors.

```rust
impl TryFrom<syn::Item> for PartialType {
    type Error = ();
}
```